### PR TITLE
Allowed unused parameters before used ones

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
This PR addresses concerns listed in #9458 by removing errors logged for any unused parameters, particularly from @KnisterPeter and myself. Errors are now skipped for unused parameters that come before used parameters.

Asking for `_`-prefixed variable names as the primary method of marking variables as unused can be non-optimal for several reasons:
* We're binding compiler logic to the variable name
* It's easy to forget to remove the prefix, such as during refactoring
* Prefixed names clash with some existing coding styles

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15958,24 +15958,57 @@ namespace ts {
...
+                }
+
+                if (compilerOptions.noUnusedLocals) {
+                    checkUnusedLocals(node.locals);
+                }
+            }
+        }
+
+        function checkUnusedParameterLocals(locals: Map<Symbol>): void {
+            let unusedParameterLocals: Symbol[] | undefined = [];
```
Comment (JoshuaKGoldberg) [inline src/compiler/checker.ts]: From a performance perspective I was hesitant to make a new [] no matter what. Hence the slight added complexity.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -15958,24 +15958,57 @@ namespace ts {
...
                         }
+                    } else if (unusedParameterLocals) {
+                        unusedParameterLocals.push(local);
+                    } else {
+                        unusedParameterLocals = [local];
                     }
                 }
             }
+
+            if (unusedParameterLocals) {
```
Comment (JoshuaKGoldberg) [inline src/compiler/checker.ts]: Should `unusedParameterLocals` be sorted? `Map` guarantees iteration in insertion order, but does TS guarantee locals be in inserted order? Do we want to rely on that guarantee?

**Comment by andy-ms** [comment]
In my experience, it's useful to know when you aren't using a parameter, regardless of where that parameter appears.
I know it can be annoying if an API wants you to provide a callback where only the last parameter is useful, but I wouldn't want a solution that doesn't consider users who want *all* of their unused parameters to be errors, not just ones on the end.

**Comment by JoshuaKGoldberg** [comment]
@andy-ms I was actually referring to the technical implementation, but that's a good point.
@mhegazy filed #12149

I'll close this PR and leave the branch as reference.

**Pull request closed without merging**